bazel: rebuild clang and fix sanitizers

rockwotj opened the pull request:
Thanks to @mmaslankaprv we've narrowed down the last of the sanitizers issue is due to the compiler. I've rebuilt clang with LTO in a cloud VM instead of locally with PGO. With the rebuilt compiler the sanitizer tests pass now :tada:

I've also upgraded to clang 19 and improved a few other things along the way.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
bazel/toolchain: upgrade to clang-19

Upgrade to the latest clang. The real motivation here is that we want
the sanitizers to work, however for some reason the custom built clang
we use has false positives with coroutines and the sanitizers.

However clang-18 is compiled against a super old version of ubuntu and
missing libraries are missing (namely `libtinfo.5.so`) and not available
on recent ubuntu and fedora distributions.

Since we're upgrading the clang-19 we need to address a few deprecations
in libc++, so we upgrade seastar for this.

rockwotj committed:
bazel: only enable batch cache test if using seastar allocator

This was incorrectly configured, the test relies on the seastar
allocator's reclaim ability, which is required to pass. As is this
always is disabled on CI but we only want to disable the test if the
system allocator is being used because the reclaim ability is not
available.

rockwotj committed:
rpc: don't move service::_methods

_methods is an array of non_copyable functions that have `[this]`
pointers captured, we cannot move them into a new object because
the `this` pointer become invalidated and pointing to the old object.
We must never move these RPC services because this would be very
buggy otherwise.

rockwotj committed:
datalake/translation: include what you use

clang 19 is more strict on this

rockwotj committed:
iobuf/fuzz: include what you use

clang 19 is more strict here

rockwotj committed:
ct: make pipeline stage id move assignable

libc++ 19 is more strict, but it requires types are copy assignable, and
according to the standard copy assignable types must be move asssignable
as well.

https://en.cppreference.com/w/cpp/named_req/CopyAssignable

rockwotj committed:
bazel: suppress leaks in openssl-fips

In getting the sanitizers work in openssl tests with fips enabled we
need to suppress leaks there. The benchmarks are full of them, and I
assume these are not caught in CMake because benchmarks don't run with
sanitizers.

rockwotj committed:
datalake: make benchmark test fast

In bazel we test benchmarks by running them once, well 1000 batches is
really slow in debug mode, so running once doesn't do anything. Instead
change the number of batches in debug mode to make the benchmark fast.

dotnwat commented on src/v/storage/tests/BUILD:
```diff
@@ -240,8 +240,8 @@ redpanda_cc_btest(
     target_compatible_with = select({
-        "@seastar//:use_system_allocator": [],
-        "//conditions:default": ["@platforms//:incompatible"],
+        "@seastar//:use_system_allocator": ["@platforms//:incompatible"],
+        "//conditions:default": [],
```
huh, i wonder why it was working before... that's concerning

rockwotj replied:
I think the default build uses the seastar allocator still, so it was just always disabled

dotnwat commented on src/v/rpc/rpc_compiler.py:
```diff
@@ -69,7 +70,7 @@ class failure_probes;
        : _sc(sc), _ssg(ssg) {}
 
     {{service_name}}_service_base({{service_name}}_service_base&& o) noexcept
-      : _sc(std::move(o._sc)), _ssg(std::move(o._ssg)), _methods(std::move(o._methods)) {}
+      : _sc(std::move(o._sc)), _ssg(std::move(o._ssg)) {}
```
> We must never move these RPC services because this would be very
buggy otherwise.

was the intention to instead delete the move consturctor?

rockwotj replied:
I'm game for deleting both the move constructor and move assignment operator - I don't think either are sound for `_methods`. I do think we don't need to move the methods and it's fine with this change.

rockwotj replied:
Ah what I was trying to say was, today in the code base there is no way we in practice move any services today, because there would be mass bugs with rpcs invoking the wrong object. Maybe we should formalize this and delete the move ctor and assignment operator

rockwotj replied:
Pushed a commit to remove them.

rockwotj commented on MODULE.bazel:
```diff
@@ -133,15 +139,17 @@ llvm.toolchain(
-        "linux-x86_64": ["http://redpanda-core-toolchain.s3-website-us-east-1.amazonaws.com/llvm-19.1.1-ubuntu-22.04-x86_64.tar.gz"],
-    },
+    llvm_version = "19.1.7",
+    # Currently our custom build of clang causes false positives with sanitizers, so we use the default clang distribution.
```
Yeah exactly, I'm not sure what. We may have to go back to custom builds and figure out what flag is calling this, because in CI we need older libraries, but locally I need newer ones

rockwotj committed:
rpc/service: remove move ctor/operator

This isn't really safe because pointers are taken to `this`, we don't
use these move operators anyways and they are a big footgun, so just
remove them.

michael-redpanda commented on lsan_suppressions.txt:
```diff
@@ -1 +1,2 @@
 leak:seastar::net::conntrack::handle::~handle
+leak:fips/self_test.c
```
Just a note:  self tests are run once per shard and, unfortunately, are within the validated section of the OpenSSL FIPS module which means we cannot touch it and maintain validation.

michael-redpanda commented:
It looks like CI cpp lint (which I assume uses clang-format-18) passes against the changes in https://github.com/redpanda-data/redpanda/pull/25120/commits/35318ed5e41b6d4807387fc6ace937babcec0371.  Do we need to ensure that core knows to use clang-format in bazel rather than in vtools?  Should we remove `task lint:cpp` in vtools to ensure it isn't accidentally used and change CI to use bazel?

rockwotj commented:
>  Should we remove task lint:cpp in vtools to ensure it isn't accidentally used and change CI to use bazel?

Yeah people should use `bazel run //tools:clang_format` now

nvartolomei commented on src/v/storage/tests/BUILD:
```diff
@@ -240,8 +240,8 @@ redpanda_cc_btest(
     target_compatible_with = select({
-        "@seastar//:use_system_allocator": [],
-        "//conditions:default": ["@platforms//:incompatible"],
+        "@seastar//:use_system_allocator": ["@platforms//:incompatible"],
```
please add a WHY to the commit messages 🙇‍♂️

rockwotj replied:
Done

michael-redpanda commented:
> > Should we remove task lint:cpp in vtools to ensure it isn't accidentally used and change CI to use bazel?
> 
> Yeah people should use `bazel run //tools:clang_format` now

Should apply as well to buildkite?

dotnwat commented:
>Should apply as well to buildkite?

yeh i think it runs as a github action?

rockwotj commented:
> > Should apply as well to buildkite?
> 
> yeh i think it runs as a github action?

https://github.com/redpanda-data/redpanda/blob/ceae5b0d0bf5e90a5083f52309b6ed1ed1314919/.github/workflows/lint-cpp.yml#L38

rockwotj committed:
bazel/toolchain: add zlib back to clang builds

This causes warnings not having it, besides we enable this in vtools and
we want parity there.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--8b1b0a6c-cf07-4972-9f66-e21cf2fdcf73-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62070">build#62070</a></summary>


| test_id                                                                                                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62070#0195252c-680c-4d5e-a96c-890949790246 | FLAKY         | 1/2      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62070#0195252c-680b-45fd-98e5-6ef8af855866 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62070#0195252c-680c-4d5e-a96c-890949790246 | FLAKY         | 1/2      |
</details>
<!--/body-->

rockwotj committed:
bazel/toolchain: use zstd compression

Because we can save some bandwidth and we normally cache this

rockwotj committed:
bazel/toolchain: don't build benchmarks

We don't need to build these, nor do we need to include llvm libraries,
we only want a toolchain (compiler, etc) here.

rockwotj committed:
bazel/toolchain: default to latest clang version

I think we can track llvm more aggressive.

Also add a sysroot for PGO builds, we'll need some of the headers
packaged in there.

rockwotj commented:
Marking as draft until I get the compiler build working with sanitizers

rockwotj committed:
bazel/toolchains: use LTO compiled clang

Switch back to using a in-house compiled LLVM toolchain.

Due to the compiler being used on a variety of linux distributions
from CI to dev workstations, we want it to be as statically linked
as possible. These binaries here statically link the C++ standard
library and are compiled in ubuntu:jammy containers so it should have
quite a bit of compatibility.

Additionally, there seems to be issues with the sanitizers in the PGO
build, so we fallback to LTO only until we track down the issue.

Lastly, we switch to github releases from an s3 bucket. The s3 bucket
is in a non-ideal project and discoverability is hard. There are also
issues like being linked to a single AZ, etc so it's slow to download
in certain places in the world. We can punt all that to Github by using
Github releases to host the release artifacts from the containers in
`bazel/toolchain/Dockerfile.*`. We already have plenty of dependencies
on Github for building and if we want to fix that we can do it in a
unified approach to mirroring our dependencies and then this naturally
becomes apart of that solution.

rockwotj committed:
bazel/toolchain: statically link libc++

This makes our toolchain more portable across distros with different
versions of the C++ stdlib.

Additionally update some PGO compiler flags to use proper settings from
the LLVM CMake files instead of directly injecting compiler flags.

rockwotj commented:
Okay with these changes I have the tests all passing with sanitizers and I've fixed our custom clang build by dropping the PGO step. I will work on re-enabling PGO in a followup PR.

rockwotj committed:
model: include what you use

dotnwat commented:
this will probably conflict with https://github.com/redpanda-data/redpanda/pull/25149/files

rockwotj commented:
Probably. I'm assuming that will finish CI first. My plan here is to get a clean CI run, then I might need to force merge to not have to constantly rebasing clang-format changes

rockwotj commented:
Hrm seems clang tidy is segfaulting :(

rockwotj committed:
bazel/toolchain: switch back to LLVM 18

Right now clang-tidy-19 crashes on our code base,
so downgrade to 18 until we fix it.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--73d1666b-9430-4bce-af9e-8e16cb1fd1a0-->
<!--TEST RESULTS-->
<!--2f444251-55da-4749-ab49-0e1f36d059b5-->
<!--TEST RESULTS-->
<!--97439036-2aff-4d61-885e-68ae75755da6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62253">build#62253</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62253#01953bd3-adb0-4050-856d-0b4fb84ac644 | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=False.controller_snapshots=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62253#01953bba-96d7-48fe-97d5-efe584eb21fc | FLAKY | 1/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62253#01953b75-8d67-4d53-a763-73ac6cc68751 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62342">build#62342</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/62342#0195437d-a629-4dde-91a0-aed8884cf266 | FLAKY | 1/2 |
| rptest.tests.partition_state_api_test.PartitionStateAPItest.test_local_summary | ducktape | https://buildkite.com/redpanda/redpanda/builds/62342#01954362-7241-4b45-930f-08dd2ce436c4 | FLAKY | 1/2 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.SegmentRolledByTimeout==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62342#0195437d-a628-496a-a0ba-c5d9fa846205 | FLAKY | 1/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62342#0195431e-2c34-44a8-92c0-37c0134e08fc | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62376">build#62376</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=None.topic_cfg=90000                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62376#0195458f-71a4-4ce0-b797-eacf3d772101 | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62376#0195457b-cbeb-4462-a899-4f225b5cdc52 | FLAKY         | 1/3      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62376#0195457b-cbeb-4462-a899-4f225b5cdc52 | FLAKY         | 1/2      |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit                                                                          | unit        | https://buildkite.com/redpanda/redpanda/builds/62376#01954533-fb20-4ed3-bcf1-0a8f0560a1c3 | FLAKY         | 1/2      |
</details>
<!--/body-->

rockwotj committed:
bazel/toolchain: always use the latest and best ABI for libc++

We package this statically in our build so there is no reason to
preserve ABI in this case and we want the highest performance ABI.

rockwotj commented:
Okay I rebuilt clang 18 and now everything is working...

I will separately upgrade to llvm 19 because we need to fixup our clang-tidy stuff first (possibly there is a bad check in clang-tidy 19, there are a few issues on github for llvm). I need to dig in more to get a stacktrace but am running out of time and I would like to cache in on working sanitizers so I can enable them in the bazel build in CI

dotnwat commented on tools/clang-format:
```diff
@@ -4,7 +4,7 @@
 # executable as possible.
 
-bazel run @llvm_18_toolchain//:clang-format \
+bazel run @llvm_19_toolchain//:clang-format \
```
18?

rockwotj replied:
Ugh thanks will fix

rockwotj replied:
Fixed

dotnwat commented on bazel/toolchain/Dockerfile.llvm:
```diff
@@ -14,7 +14,7 @@ ADD --chmod=755 https://github.com/kadwanev/retry/raw/20997c7712a4/retry /usr/bi
 WORKDIR /opt
 
-ARG LLVM_VERSION=19.1.6
+ARG LLVM_VERSION=19.1.7
```
so is this being overridden, or its irrelevant because are using pre-built v18?

rockwotj replied:
I've been overriding it when I build because I'm testing many different versions

rockwotj commented:
Oops, accidentally rebased. I only changed `s/9/8` in `tools/clang_format`

dotnwat reviewed (approved):
no complaint here!

nvartolomei commented on src/v/rpc/rpc_compiler.py:
```diff
@@ -68,16 +69,8 @@ class failure_probes;
-       }
-       return *this;
-    }
+    {{service_name}}_service_base({{service_name}}_service_base&& o) noexcept = delete;
```
This could have been left but made ~~private~~ protected instead? I.e. derived classes if needed could implement move constructor by first calling this one and avoiding having to deal with _sc, _ssg, _methods. Right?

Hypothetical question.

rockwotj replied:
could have I suppose, but it is still sketchy

nvartolomei replied:
What exactly is sketchy?

dotnwat replied:
pretty sure tyler is referring to how the service base type stores an array of lambdas that capture `this` for each handler. if moves were allowed, then all that would have to be handled. so given that no user currently moves, its a conservative approach?

rockwotj replied:
Sorry yes @dotnwat said it better than I did. Do you have a use case where you want to move a service? IIRC we always allocate them in a `std::unique_ptr` today so it seems simpler to just not worry about any of this?

dotnwat replied:
>always allocate them in a std::unique_ptr today so it seems simpler to just not worry about any of this?

good point. that's ideal

nvartolomei replied:
> pretty sure tyler is referring to how the service base type stores an array of lambdas that capture this for each handler.

Yikes. Great we explicitly deleted it then.

> Do you have a use case where you want to move a service?

I don't. Came across this PR as it conflicted with a change I was doing (making the class) where I also had to remove the copy constructor as `new base` didn't make sense anymore. First I removed the constructors but then I thought about making them protected just in case they will be needed in the future. Wanted to understand your thought process for arriving at a "different" end state (for move constructor, move assignment had to go anyway).

Thank you both for insights!
